Java equality between chars fails if one is unboxed.

### Dafny version

latest-nightly

### Code to produce this issue

```dafny
predicate TestX(f: char -> bool) {
  f('x')
}

predicate TestXOn(obtained: char) {
  TestX((c: char) => c == obtained)
}

method Main() {
  var c := TestXOn('x');
  print c;
}
```

### Command to run and resulting output

```code
> dafny run --target=java

Dafny program verifier finished with 0 verified, 0 errors
Picked up JAVA_TOOL_OPTIONS: -Dlog4j2.formatMsgNoLookups=true
IssueChar-java\_System\__default.java:15: error: bad operand types for binary operator '=='
      return (_1_c) == (_0_obtained);
                    ^
  first type:  int
  second type: CodePoint
IssueChar-java\_System\__default.java:16: error: incompatible types: int cannot be converted to CodePoint
    })).apply(obtained));
              ^
Note: Some messages have been simplified; recompile with -Xdiags:verbose to get full output
2 errors
Error while compiling Java files. Process exited with exit code 1
```

### What happened?

The code did not compile in Java. I caught this bug while trying to make the Parsers library to pass.

### What type of operating system are you experiencing the problem on?

Windows

--- a/Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs
+++ b/Source/DafnyCore/Backends/Java/JavaCodeGenerator.cs
@@ -3792,10 +3792,25 @@ namespace Microsoft.Dafny.Compilers {
       List<Type> boundTypes, Type resultType, IOrigin resultTok, bool inLetExprBody, ConcreteSyntaxTree wr,
       ref ConcreteSyntaxTree wStmts) {
       wr.Write("(({0}<{1}{2}>)", DafnyFunctionIface(boundTypes.Count), Util.Comma("", boundTypes, t => BoxedTypeName(t, wr, resultTok) + ", "), BoxedTypeName(resultType, wr, resultTok));
-      wr.Write($"({Util.Comma(boundVars)}) -> ");
+      var unboxStatements = new ConcreteSyntaxTree();
+      var realBoundVars = boundVars.Zip(boundTypes,
+        (boundVar, type) => {
+          if (BoxedTypeName(type, wr, resultTok) == TypeName(type, wr, resultTok)) {
+            return boundVar;
+          }
+
+          var newName = ProtectedFreshId("_" + boundVar);
+          var rhsWriter = DeclareLocalVar(boundVar, type, resultTok, unboxStatements);
+          rhsWriter = EmitCoercionIfNecessary(NativeObjectType, type, resultTok, rhsWriter);
+          rhsWriter.Write(newName);
+          return newName;
+        }).ToList();
+      wr.Write($"({Util.Comma(realBoundVars)}) -> {{");
+      wr.Append(unboxStatements);
+      wr.Write("return ");
       var w = wr.Fork();
-      wr.Write(").apply");
-      TrExprList(arguments, wr, inLetExprBody, wStmts);
+      wr.Write(";}).apply");
+      TrExprList(arguments, wr, inLetExprBody, wStmts, i => NativeObjectType);
       return w;
     }
 

--- a/Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs
+++ b/Source/DafnyCore/Backends/SinglePassCodeGenerator/SinglePassCodeGenerator.cs
@@ -5167,7 +5167,9 @@ namespace Microsoft.Dafny.Compilers {
       return wr;
     }
 
-    void CreateFreeVarSubstitution(Expression expr, out List<BoundVar> bvars, out List<Expression> fexprs, out Substituter su) {
+    protected void CreateFreeVarSubstitution(
+      Expression expr,
+      out List<BoundVar> bvars, out List<Expression> fexprs, out Substituter su) {
       Contract.Requires(expr != null);
 
       var fvs = FreeVariablesUtil.ComputeFreeVariables(Options, expr);

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy
@@ -1,0 +1,14 @@
+// RUN: %testDafnyForEachCompiler "%s"
+
+predicate TestX(f: char -> bool) {
+  f('x')
+}
+
+predicate TestXOn(obtained: char) {
+  TestX((c: char) => c == obtained)
+}
+
+method Main() {
+  var c := TestXOn('x');
+  print c;
+}

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy
@@ -10,5 +10,6 @@ predicate TestXOn(obtained: char) {
 
 method Main() {
   var c := TestXOn('x');
-  print c;
+  var d := TestXOn('y');
+  print c, d;
 }

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
@@ -1,0 +1,2 @@
+
+Dafny program verifier finished with TODO verified, TODO errors

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
@@ -1,3 +1,1 @@
-
-Dafny program verifier finished with 0 verified, 0 errors
 truefalse

--- a/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
+++ b/Source/IntegrationTests/TestFiles/LitTests/LitTest/git-issues/git-issue-6149.dfy.expect
@@ -1,2 +1,3 @@
 
-Dafny program verifier finished with TODO verified, TODO errors
+Dafny program verifier finished with 0 verified, 0 errors
+truefalse

--- a/docs/dev/news/6149.fix
+++ b/docs/dev/news/6149.fix
@@ -1,0 +1,1 @@
+Java lambdas now accepting boxing/unboxing

